### ISSUE TITLE
Crypto.KDF.BCrypt docs say the latest version is 2b, but 2a is generated

### ISSUE THREAD

**Issue by MaxGabriel** [haskell-crypto/cryptonite#203]
The docs for [Crypto.KDF.BCrypt](https://hackage.haskell.org/package/cryptonite-0.24/docs/Crypto-KDF-BCrypt.html) say:

> The different version numbers have evolved because of bugs in the standard C implementations. **The most up to date version is 2b and this implementation the 2b version prefix**, but will also attempt to validate against hashes with versions 2a and 2y. Version 2 or 2x will be rejected. No attempt is made to differentiate between the different versions when validating a password, but in practice this shouldn't cause any problems if passwords are UTF-8 encoded (which they should be).

This suggested to me that if I generated a bcrypt hash with the latest version of the package, I would get a 2b hash, but when I generate a hash I get the 2a version:

```
import Crypto.KDF.BCrypt
import qualified Data.ByteString.Char8 as B
let password = B.pack "password"
hashPassword 12 password :: IO B.ByteString
"$2a$12$i9vPMbmOjH4lDaW0DAmG6uZTza1gbRDlxHyOnMn/ek/mcSMhS0.Um
```

Is this a bug, or am I just misreading the docs here?

I've tested against versions 0.23 and 0.24.

**Comment by tekul** [comment]
Yeah, that is a bit confusing - it reads as if the code will output hashes using a "2b" version prefix. In practice the different versions are only there to account for specific bugs in the Open BSD code, so it shouldn't matter. I think I left it at "2a" for backwards compatibility with other implementations which might not have recognized the "2b" version, since it was [only introduced the year before](https://marc.info/?l=openbsd-misc&m=139320023202696). For example, python's passlib didn't support "2b" until [a year after this code was written](https://passlib.readthedocs.io/en/stable/history/1.7.html#id3).

As the doc says, this implementation makes no attempt to differentiate between the versions, so if you have an old hash which was created with one of those buggy implementations (and which falls into one of the cases they cater for) then validating against it may fail, but the "2b" vs "2a" bug only applies to passwords which are longer than 256 characters, which is an unlikely corner case for most people.

I think it's safe to use "2b" from now on (other implementations should support that by now) so I'll modify the code to be consistent with the docs. Thanks for the report!

Correction: passlib [did already support "2b"](https://passlib.readthedocs.io/en/stable/history/1.6.html#id2), before then. It just made it the default in version 1.7.

**Comment by tekul** [comment]
Fixed in  5e354f9bfcf3.

**Comment by MaxGabriel** [comment]
Got it, thanks @tekul!
